lib/unlock-helpers.js

```diff
@@ -131,6 +135,28 @@ helpers.pinUnlock = async function pinUnlock (adb, driver, capabilities) {
   }
 };
 
+helpers.pinUnlockWithKeyEvent = async function pinUnlockWithKeyEvent (adb, driver, capabilities) {
+  logger.info(`Trying to unlock device using pin with keycode ${capabilities.unlockKey}`);
+  await helpers.dismissKeyguard(driver, adb);
+  const keys = helpers.stringKeyToArr(capabilities.unlockKey);
+
+  // Some device does not have system key ids like 'com.android.keyguard:id/key'
+  // Then, sending keyevents are more reliable to unlock the screen.
+  for (const pin of keys) {
+    // 'pin' is number (0-9) in string.
+    // Number '0' is keycode '7'. number '9' is keycode '16'.
+    await adb.shell(['input', 'keyevent', parseInt(pin, 10) + NUMBER_ZERO_KEYCODE]);
+  }
+
+  // Some devices accept entering the code without pressing the Enter key
+  // When I rushed commands without this wait before pressKeyCode, rarely UI2 sever crashed
+  await sleep(UNLOCK_WAIT_TIME);
```

Comment:
can we extract this duplicated block into a separate function to avoid duplication?

Reply:
👍